8344637: Fix Page8 of manual test java/awt/print/PrinterJob/PrintTextTest.java on Linux and Windows

gredler opened the pull request:
This PR fixes the issue identified in JDK-8148334 in screenshots `Page8_landscape.JPG` and `Page8_portrait.JPG`.

It does not address `mac_Page1.png` or `mac_Page8.png`, which I'm not even sure are still issues (I have no access to a Mac).

The method in question, `PathGraphics.printedSimpleGlyphVector(...)` is quite complex, with many special cases being handled in different ways. In this specific scenario (page 8 of `PrintTextTest`), all special case checks fail, and we fall through all the way to the final handling block, which draws the individual characters one by one. It looks like the problem is that the font transform translation is applied twice, once via the glyph positions, and again by `drawString(...)` via the font. The proposed fix is to provide `drawString(...)` a font without any translation transform.

Testing looks good on Linux, but needs to be done on Mac and Windows.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8344637](https://bugs.openjdk.org/browse/JDK-8344637): Fix Page8 of manual test java/awt/print/PrinterJob/PrintTextTest.java on Linux and Windows (**Bug** - P3)


### Reviewers
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21980/head:pull/21980` \
`$ git checkout pull/21980`

Update a local copy of the PR: \
`$ git checkout pull/21980` \
`$ git pull https://git.openjdk.org/jdk.git pull/21980/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21980`

View PR using the GUI difftool: \
`$ git pr show -t 21980`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21980.diff">https://git.openjdk.org/jdk/pull/21980.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21980#issuecomment-2464948928)
</details>

gredler commented:
@honkar-jdk This is the followup to PR #21716 discussed last week. Do you have time to review and possibly test on Windows or mac? The change looks good for me on Linux.

honkar-jdk commented:
> @honkar-jdk This is the followup to PR https://github.com/openjdk/jdk/pull/21716 discussed last week. Do you have time to review and possibly test on Windows or mac? The change looks good for me on Linux.

Sure. I can check it on windows & macOS and get back to you.

honkar-jdk commented:
@gredler  Tested on windows with print to pdf option. I noticed that the "G..y..h..." line is still being printed above instead of below as seen in the following screenshot. (Left one - without changes, center window - test window, Right one- with the fix). 

If you could attach a similar screenshot of your testing on linux, it would be easier to review.

<img width="1403" alt="GlyphPrint" src="https://github.com/user-attachments/assets/ce43785e-439b-4d1e-ad48-8fd7101fcb81">

gredler commented:
@honkar-jdk Interesting, it looks like this fixes Linux then, but not Windows (although the code change is in the shared `PathGraphics` class). Any idea what Mac looks like on page 8 right now? It looks to me like the Windows screenshot is showing two separate bugs (start position, and then individual glyph advances) -- but that's just a guess.

We may need to create a separate bug in the bug tracker just for Linux page 8 and this PR, since it looks like there are still other issues to be fixed in JDK-8148334.

Linux screenshot:

![linux](https://github.com/user-attachments/assets/61187b29-ce74-497d-a5be-bc6ba1d40e41)

DamonGuy commented:
@gredler MacOS looks different. There seems to be some clipping issues or bound issues when printing. In fact, I don't even see the `GlyphVector` characters being printed. Could not be printed at all or moved somewhere off the page. Really odd. The changes made here don't seem to affect this though since it seems to be an issue both before & after.

![image](https://github.com/user-attachments/assets/f62671e2-85ef-482a-be9c-aa077982ed5a)

gredler commented:
@DamonGuy Brilliant, thanks for the check. This feels like it might be drawing offscreen because of incorrectly-combined transforms, like what was fixed in PR #20993 (but this would be somewhere in the Mac-specific code).

@honkar-jdk I was able to dig up an old Windows laptop, and have confirmed that the page 8 issue on Windows is caused by two regressions introduced over the past decade but not caught because this is a manual test, not an automated test... I'm going to see if I can't get this test passing on both Linux and Windows. Keep an eye out for additional commits to this PR.

honkar-jdk commented:
@gredler Checked on windows with the latest fix. The portrait looks good but looks like the fix causes more issues in landscape mode. I require some time to go through the source code changes. Looks like you might be in the process of adding few more changes for Windows issue. Let me know once the PR is ready.

**Portrait**

<img width="1393" alt="Win_PortraitPg#8" src="https://github.com/user-attachments/assets/a6b48dfe-19f0-4739-825f-e50b0ae50128">

**Landscape**

<img width="1532" alt="Win_LandscapePg#8" src="https://github.com/user-attachments/assets/eeeb309f-fd36-4b8f-a77b-d53acc8a178b">

gredler commented:
@honkar-jdk OK, I think this PR is ready for another review on Windows. The issue you highlighted in your last comment was caused by a partial fix to the JDK-8132890 regression (which had obviously removed the font rotation component, but had less-obviously also removed the device transform rotation component).

The goal is to completely fix this test on Windows and Linux, so I've updated the problem list to flag only macOS as problematic. Will we need a separate bug in the issue tracker, since this PR will not fix the macOS parts of JDK-8148334? I'm not sure whether PRs are allowed to partially fix issues (when they represent more than one bug), or whether they are expected to fully fix whatever issue they reference.

@prrace This is related to two bugs that you worked on in the past: JDK-8132890 and JDK-8029204. It has been a few years, but it would be good to get your review on this as well, if you have the time.

In particular, it feels like we stopped using most of the device transform in some parts of the JDK-8132890 changes (e.g. glyph position calculation), and now we're re-adding the rotation component, and I wonder if there are other parts of the device transform that we are missing, and/or whether it would be easier to go back to using the device transform in the glyph position calculation. (The code seems to work as-is, I'm just wondering if it's more complex than is necessary.)

Summary of changes:

* PathGraphics: Do not apply font transform translation component twice (results in incorrect text start point)
* WPathGraphics: Do not apply font transform translation component twice (results in incorrect text start point)
* WPathGraphics: Include both font and device rotation when calculating glyph positions
* WPathGraphics: Fix direction of y scale

honkar-jdk commented:
@gredler 

> Will we need a separate bug in the issue tracker, since this PR will not fix the macOS parts of JDK-8148334? 

That's correct. I think it is better to create a new JBS bug and use the new JBS ID for this PR than using [JDK-8148334](https://bugs.openjdk.org/browse/JDK-8148334). Let me know if you need any help with creating a new JBS issue. 
Since this test span across multiple pages testing various glyph transformations it is better to keep [JDK-8148334](https://bugs.openjdk.org/browse/JDK-8148334) in open state and reference it while problemlisting the test.

> I'm not sure whether PRs are allowed to partially fix issues (when they represent more than one bug), or whether they are expected to fully fix whatever issue they reference.


A PR can have fix that focuses on specific issue and separate PR can be created to address other issues. So in this case it seems reasonable to have a PR for Windows and Linux fix and create a separate one for macOS.

gredler commented:
@honkar-jdk I just got access to JBS today, so that's not a problem. I've created JDK-8344637 to track just the Linux and Windows fixes in this PR (scoping out macOS for now). Have you had a chance to review on Windows?

honkar-jdk commented:
@gredler 
>  Have you had a chance to review on Windows?

Yes, following are the screenshots of Page8 on Windows. Both portrait and landscape look good now although I haven't looked at the other pages in detail (to verify the fix works in all cases and doesnot produce any side-effects/regressions).

<img width="1388" alt="Latest_Pg#8Portrait" src="https://github.com/user-attachments/assets/d6ee00a9-e16d-4405-bc22-9d1f6b09e66b">

<img width="1393" alt="Latest_Pg#8Landscape" src="https://github.com/user-attachments/assets/89dc1734-8ea4-4713-aa1f-c20c892f2734">

honkar-jdk commented:
@gredler The PR title needs to be updated to the latest JBS title ( I added Page8 to be more specific).

gredler commented:
@honkar-jdk OK, I've updated the title. I've checked output on both Windows and Linux and I don't think there are any other issues with this test on either platform.

honkar-jdk commented on src/java.desktop/share/classes/sun/print/PathGraphics.java:
```diff
@@ -934,23 +934,36 @@ boolean printedSimpleGlyphVector(GlyphVector g, float x, float y) {
             return true;
         }
 
+        Font font2 = font;
+        if (font.isTransformed()) {
+            AffineTransform t = font.getTransform();
+            if ((t.getType() & AffineTransform.TYPE_TRANSLATION) != 0) {
+                t.setTransform(t.getScaleX(), t.getShearY(), t.getShearX(), t.getScaleY(), 0, 0);
+                font2 = font.deriveFont(t);
+            }
+        }
+
```
The fix to PathGraphics.java looks good.

A new font object is created where you copy over all the transforms except translation from the original font object and use the newly created font object in drawString(). The translation transform is skipped since the drawString takes care of x and y positions.

honkar-jdk commented on src/java.desktop/windows/classes/sun/awt/windows/WPathGraphics.java:
```diff
@@ -876,6 +869,8 @@ private void textOut(String str,
               */
              AffineTransform advanceTransform =
                 AffineTransform.getScaleInstance(scaleFactorX, scaleFactorY);
+             advanceTransform.rotate(devangle); // radians
+             advanceTransform.rotate(iangle * Math.PI / 1800.0); // 1/10 degrees -> radians
```
Fix to WPathGraphics.java is a little more intricate and certainly requires a more expert review from @prrace 
Was devangle fix added for the landscape issue observed on Pg#8 ?

A earlier comment mentions that device space rotation must not be added. 

```
 advanceTransform.rotate(devangle); // radians
 advanceTransform.rotate(iangle * Math.PI / 1800.0); // 1/10 degrees -> radians
```

gredler replied:
The two angles (`iangle` + `devangle`) need to be used to fix the Windows page 8 printed "GlyphVector with position adjustments" test case, here's what it looks like without these rotation angles (left side is printed and incorrect, right side is not printed and is correct):

![no-angle](https://github.com/user-attachments/assets/0366606d-b71c-4a4c-9d2c-402144aaf7d3)

Adding `iangle` alone looks OK in the printed portrait page, but not in the printed landscape page (`devangle` is zero and therefore makes no difference when printing in portrait mode, but not so for landscape mode).

This restores the rotation transformation that was being applied prior to commit 5c2672351683467fdeeda7796d971634fc3c51ae (prior to this commit the advance transform was a copy of the device transform, with the extra `iangle` rotation applied, meaning that both the device rotation and the `iangle` rotation were included).

honkar-jdk replied:
Thanks for including the details. Sounds reasonable.

honkar-jdk reviewed (commented):
Pg#8 looks good. I did not notice any regressions on other pages with the fix on windows.

honkar-jdk commented:
/reviewers 2 reviewer

honkar-jdk reviewed (approved):
Testing looks good on Windows and no regressions are observed on other pages.

gredler commented:
@prrace Might you have time to review this PR in the next week or two, now that the SecurityManager exorcism is complete? The code being changed is related to a few bug fixes that you implemented over the past few years, so I'm not sure if anybody else really has as much context on this one as you do. Thanks!

gredler commented:
Bot, please keep open.

@prrace Friendly reminder that this one probably needs your review.

prrace reviewed (approved):
I've tested the fix on Mac, Linux and Windows.
It is painstaking comparing all the pages on-screen and off-screen to be sure not to overlook something.
I had to pause the test timer to review it all before the test timed out.

There are a couple of before and after things to note (ie pre-existing problems, which remain an issue)
Eg on Linux Page 3, "drawChars(char[]" prints as "drawChars(   char[]"
I don't know where the extra space came from.
but I should note I am viewing this in a Postscript viewer, not on a printer and it is the same before and after.

On windows, pages 12 / 13 - using the Japanese font - on screen uses embedded bitmaps so
the size doesn't scale smoothly line to line. It 'jumps' when the next bitmap size is selected.
The printing case size increases smoothly because bitmaps aren't used there.
This is a small, pre-existing anomaly and the instructions have a comment that text does not
scale linearly screen to printer, so that is not necessarily a bug and the on-screen case
is actually what you might want to change, but not at the cost of losing the bitmaps ..
But the difference is pretty extreme in this case. We may need to re-visit this.

macOS is still a problem.
Notably every page has an issue with the glyphvector positioning case being wildly wrong.
But again, pre-existing. I didn't see anything  get worse with this change.

So I'm approving the fix, although the *pre-existing* problems on windows and linux may mean
that some tester might consider the anomalies there worth a new bug report.

gredler commented:
/integrate

prrace commented:
/sponsor
